### PR TITLE
#167313410 Users should be able to receive notifications

### PR DESCRIPTION
**Pull request by nkalyesubula**
#### What does this PR do?

-  It adds a feature for enabling users to receive notifications

#### Description of Task to be completed?

- **As a User**
    I want to be able to receive notifications when users I follow publish new articles.
    I want to be able to receive notifications when articles I favorited have new interactions (like 
    new comments).
    I want to be able to opt-in and out of these notifications.

#### How should this be manually tested?

- clone the repo, and then run 
   1.  **npm run migration**
   2. **npm run seeds**
   3. **npm run dev**
   

#### What are the relevant pivotal tracker stories?
[#167313410](https://www.pivotaltracker.com/story/show/167313410)
#### Screenshots (if appropriate)
Screen Shot 2019-08-22 at 18 38 42
Screen Shot 2019-08-22 at 18 39 03
Screen Shot 2019-08-22 at 18 39 51
Screen Shot 2019-08-22 at 18 40 06
Screen Shot 2019-08-22 at 18 40 33
Screen Shot 2019-08-22 at 18 40 43
Screen Shot 2019-08-23 at 10 37 42
Screen Shot 2019-08-23 at 10 31 36
Screen Shot 2019-08-23 at 10 30 52
Screen Shot 2019-08-23 at 10 38 05

### CODE REVIEW

**Comment by Mnickii** [comment]
rebase @nkalyesubula

File: package.json [hunk 1]
```diff
@@ -40,13 +40,15 @@
     "lodash": "^4.17.15",
     "method-override": "^2.3.10",
     "multer": "^1.4.2",
+    "nodemailer": "^6.3.0",
```
Comment (Mnickii) [inline package.json]: we're not using this for now right?

**Comment by nkalyesubula** [comment]
> rebase @nkalyesubula

Rebased, and merge conflits are gone.

File: sequelize-data.json [hunk 2]
```diff
@@ -1,3 +0,0 @@
-[
```
Comment (Mnickii) [inline sequelize-data.json:1]: we dont need this anymore?

**Review by Mnickii — APPROVED** [review]

File: src/routes/api/article/article.routes.js [hunk 3]
```diff
@@ -5,11 +5,13 @@ import imageUpload from '../../../middlewares/multer';
...
 import { schema } from '../../../middlewares/validators/schemas';
 import confirmEmailAuth from '../../../middlewares/emailVarification.middleware';
+import validateId from '../../../middlewares/validators/articleId.validation';
+import FavoritesController from '../../../controllers/favorited.articles.controller';
 import { checkQuery } from '../../../middlewares/query.check';
 
 const router = express.Router();
 
-
+router.post('/:articleId/favorite', [auth, validateId], FavoritesController.createOrRemoveFavorite);
```
Comment (minega25) [inline src/routes/api/article/article.routes.js]: You can add the verifyEmail middleware

File: src/controllers/comments.controller.js [hunk 4]
```diff
@@ -35,6 +38,7 @@ class Comments {
         parentCommentId: req.body.parentCommentId,
       };
       const createdComment = await commentsService.addComment(comment);
+      notifyUsersWhoFavorited(req, res, getArticle.id, req.params.slug);
```
Comment (minega25) [inline src/controllers/comments.controller.js]: This function is sending an email as well. You can add await keyword in order to be consistent

**Review by minega25 — CHANGES_REQUESTED** [review]
Great job. I have included certain areas for improvement that can be of help in the future

File: package.json (hunk 1 above)
Comment (nkalyesubula) [reply package.json]: Fixed, please check again

File: sequelize-data.json (hunk 2 above)
Comment (nkalyesubula) [reply sequelize-data.json:1]: Fixed, please check again

File: src/controllers/comments.controller.js (hunk 4 above)
Comment (nkalyesubula) [reply src/controllers/comments.controller.js]: Fixed, please check again

**Review by minega25 — APPROVED** [review]
LGTM

File: src/controllers/user.controller.js [hunk 5]
```diff
@@ -1,3 +1,4 @@
+import nodemailer from 'nodemailer';
```
[the remark this answers is not in the document]
Comment (Mnickii) [reply src/controllers/user.controller.js]: we removed this module @nkalyesubula

**Review by Mnickii — CHANGES_REQUESTED** [review]

**Review by nkalyesubula — COMMENTED** [review]
@Mnickii fixed

**Review by salviosage — COMMENTED** [review]
Nike work @nkalyesubula, however, in my opinion, I would suggest if you use article slug instead of ids in routes.

File: sequelize-data.json (hunk 2 above)
Comment (Mnickii) [reply sequelize-data.json:1]: why are we removing this?

**Review by Mnickii — COMMENTED** [review]
@nkalyesubula

**Review by habinezadalvan — APPROVED** [review]

**Review by Mnickii — APPROVED** [review]
LGTM

**Pull request merged**